src/virtio/net.c

```diff
@@ -0,0 +1,406 @@
+{
+    struct virtio_net_device *state = device_state(dev);
+
+    if (net_queue_full_active(&state->tx)) {
+        LOG_NET_WRN("TX used ring full\n");
+        goto fail;
+    }
+
+    net_buff_desc_t sddf_buffer;
+    int error = net_dequeue_free(&state->tx, &sddf_buffer);
+    if (error) {
+        LOG_NET_WRN("Avail ring is empty\n");
+        goto fail;
+    }
+
+    void *dest_buf = state->tx_data + sddf_buffer.io_or_offset;
+
+    uint32_t written = 0;
+    uint32_t dest_remaining = NET_BUFFER_SIZE;
+
+    /* Strip virtio header before copying to sDDF */
+    uint32_t skip_remaining = sizeof(struct virtio_net_hdr_mrg_rxbuf);
+
+    struct virtq_desc *desc = &virtq->desc[desc_head];
+
+    while (dest_remaining > 0) {
+        uint32_t skipping = 0;
+        /* Work out how much of this descriptor must be skipped */
+        skipping = MIN(skip_remaining, desc->len);
+        /* Truncate packets that are large than BUF_SIZE */
+        uint32_t writing = MIN(dest_remaining, desc->len - skipping);
+
+        memcpy(dest_buf + written, (void *)desc->addr + skipping, writing);
+
+        skip_remaining -= skipping;
+        written += writing;
+        dest_remaining -= writing;
+
+        if (desc->flags & VIRTQ_DESC_F_NEXT) {
+            desc = &virtq->desc[desc->next];
+        } else {
+            break;
+        }
+    }
+
+    sddf_buffer.len = written;
+    error = net_enqueue_active(&state->tx, sddf_buffer);
+    /* This cannot fail as we check above */
+    assert(!error);
+
+    virtq_enqueue_used(virtq, desc_head, written);
+    *respond_to_client = true;
+    *notify_tx_server = true;
+    return;
+
+fail:
+    virtq_enqueue_used(virtq, desc_head, 0);
+    *respond_to_client = true;
+}
+
+static int virtio_net_queue_notify(struct virtio_device *dev)
+{
+    struct virtio_net_device *state = device_state(dev);
+
+    if (!driver_ok(dev)) {
+        LOG_NET_ERR("Driver not ready\n");
+        return 0;
+    }
+    if (dev->data.QueueSel != VIRTIO_NET_TX_VIRTQ) {
+        LOG_NET_ERR("Invalid queue\n");
+        return 0;
+    }
+    if (!dev->vqs[VIRTIO_NET_TX_VIRTQ].ready) {
+        LOG_NET_ERR("TX virtq not ready\n");
+        return 0;
+    }
+
+    virtio_queue_handler_t *vq = &dev->vqs[VIRTIO_NET_TX_VIRTQ];
+    struct virtq *virtq = &vq->virtq;
+
+    uint16_t guest_idx = virtq->avail->idx;
+    uint16_t idx = vq->last_idx;
+
+    bool notify_tx_server = false;
+    bool respond_to_client = false;
+
+    for (; idx != guest_idx; idx++) {
+        uint16_t desc_head = virtq->avail->ring[idx % virtq->num];
+        handle_tx_msg(dev, virtq, desc_head, &notify_tx_server, &respond_to_client);
+    }
+
+    vq->last_idx = idx;
+
+    if (notify_tx_server && net_require_signal_active(&state->tx)) {
+        net_cancel_signal_active(&state->tx);
+        microkit_notify(state->tx_ch);
+    }
+    if (respond_to_client) {
+        if (!virtio_net_respond(dev)) {
+            return 0;
+        }
```

Comment:
Shouldn't this just be `return virtio_net_respond(dev)`? Wouldn't returning 0 mean that we do not have an error?

Reply:
It seems all mmio devices currently use 1 for success and 0 for failure, we should probably change

Reply:
We should change all of them to use bool, I can do that now.

Reply:
I've merged https://github.com/au-ts/libvmm/pull/96 now.